<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>OMICS MEET ML</title>
    <meta charset="utf-8" />
    <meta name="author" content="Pietro Franceschi" />
    <meta name="date" content="2022-02-21" />
    <script src="libs/header-attrs-2.10/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# OMICS MEET ML
## Things to take care of
### Pietro Franceschi
### Unit of Computational Biology - Fondazione E. Mach
### 2022-02-21

---









class: inverse, left, top
# My Scientific Id

- **PhD** in Physics

--

- **Research Interests**: analysis of complex data, method development, metabolomics, mass spectrometry

--

- **Tools**: R and (less often) Python

--

- **Other Interests**: clarinet playing, bonsai, aquarium, DIY in general, ...

---
class: inverse, center, middle
# Omics

---
# What are omics

Adding the suffix _omic_ to a word is a way to indicate the **desire** of performing an **holistic large scale investigation** of a specific subject 

--

* genomics (and meta-genomics)

--

* proteomics

--

* metabolomics (with its extensions like .red[lipidomics], gl ycomics, ...)

--

* ...

--

.footnote[
The rise of _quantitative_ technologies are transforming almost all disciplines into _omics_ ... [tholinomics](https://pubmed.ncbi.nlm.nih.gov/20102174/), [petroleomics](https://www.pnas.org/content/105/47/18090)
]

---
class: inverse, left, middle
# Common Ideas

--

* challenging

--

* comprehensive and holistic (as much as possible)

--

* data rich (measuring a large bunch of variables)

--

* complex (data processing and interpretation require bioinformatics)

--

* multidisciplinary (nobody can do everything alone)


---
class: center, middle
# The role of bioinformatics/biostatitics

Spot/show a result present in data

--

Show/assess that my result holds for all the population ... **A scientific result have to be general**!

--

Allow me to reproduce my results

---
class: inverse,center,middle

## What is a **result**? A result is **organization** !

---
class: center
# A biomarker! :-)

![](omics_meet_ML_files/figure-html/unnamed-chunk-2-1.png)&lt;!-- --&gt;


---
class: center
# A non-biomarker! :-(

![](omics_meet_ML_files/figure-html/unnamed-chunk-3-1.png)&lt;!-- --&gt;

---
class: center
# Correlated variables! :-)

![](omics_meet_ML_files/figure-html/unnamed-chunk-4-1.png)&lt;!-- --&gt;


---
class: center
# Uncorrelated variables ... :-(

![](omics_meet_ML_files/figure-html/unnamed-chunk-5-1.png)&lt;!-- --&gt;



---
class: left, middle
# Data analysis goals

--

* Visualize and spot these results in my data matrix (**Exploratory Data Analysis**) 

--

* Generalize: at least *give a confidence* to my desire of generality (**statistical analysis and ML**)

---
class: left, middle
# Data deluge

* untargeted metabolomics: 10000 variables per sample
* targeted metabolomics: 300 metabolites per sample
* proteomics: 3000 proteins per sample
* metagenomics: 100000 OTUs
* NGS: many, many
* phenotyping with sensors (10 variables, every day/every half an hour)

---
class: left, middle
# Data matrices

Analogous of an Excel table

* rows are samples
* columns are variables

--

By definition, in *omics* the number of variables largely exceed the number of samples, and _technological development worsen_ this unbalancing

---
class: inverse, center, middle
# Characteristics of omics data you should always remember

---
class: center, middle
# Sample to variable unbalancing

In a typical _omic_ experiment the number of variables you measure largely exceeds the number of samples

It is **not** unlikely that the organization you measure is there only by chance

This is the result of **sampling**

.note[As we will see, the chance of finding random organization grows with the number of variables we measure]


---
class: center, middle
# Presence of unknown sub populations

The omic technology you are using to investigate your population will be able to discover **unexpected** and **hidden** structure of your sample

You are looking to your samples with a sort of _augmented reality_ device

Uniform groups are not anymore uniform !


---
class: center, middle
# Multi level experimental designs

With the increasing availability of low cost _omic_ technologies, we are able to apply such investigations to **multilevel experimental designs**

--

**Samples are not independent!**

--

Ex1: Profiling the urine of one individual over time

--

Ex2: Study how the "lake" metagenome is changing with the depth over a set of different alpine lakes

--

### Question
**How can I take this aspects into account when analyzing my data?**


---
class: center, middle
# Batch effects

![](batch.png)

.footnote[https://slfan2013.github.io/SERRF-online/#]

---
class: center, middle
# High Dynamic Range

A technology with **high dynamic range** is able to measure quantities over a large range of intensities/abundances

--

In other words: you measure together things that are abundant and things that are rare

--

E.g. In metabolomics concentrations can vary over 6 order of magnitude

--

### Questions

Should we scale?
Is reliability an issue?


---
class: center, middle
# Missing Values

**Missing values** are holes in my data matrix ... remember that **0 is not missing! Zero is zero**


--

### They arise

Errors (somewhere)

Low intensities/abundances

--

### Beware!

If we "fill in" missing values with the wrong data we bias our analysis 


---
class: center, middle
# Missing Values: questions

Should I always fill them?

--

What number should I put there?


--

How can I be sure that my choice was good?



    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
